stdlib/bytes.ml

```diff
@@ -322,6 +322,50 @@ type t = bytes
 let compare (x: t) (y: t) = Stdlib.compare x y
 external equal : t -> t -> bool = "caml_bytes_equal" [@@noalloc]
 
+let char_hex n =
+  Char.unsafe_chr (n + if n < 10 then Char.code '0' else (Char.code 'a' - 10))
+
+let to_hex s =
+  let l = length s in
+  let result = create (l * 2) in
+  for i = 0 to (l - 1) do
+    let x = Char.code (unsafe_get s i) in
+    unsafe_set result (i * 2) (char_hex (x lsr 4));
+    unsafe_set result (i * 2 + 1) (char_hex (x land 0x0f));
+  done;
+  result
+
+exception Invalid of int
+
+let from_hex_gen b =
+  let l = length b in
+  if l mod 2 <> 0 then invalid_arg "String.from_hex";
+  let exception E in
+  let digit c =
+    match c with
+    | '0'..'9' -> Char.code c - Char.code '0'
+    | 'A'..'F' -> Char.code c - Char.code 'A' + 10
+    | 'a'..'f' -> Char.code c - Char.code 'a' + 10
+    | _ -> raise E
+  in
+  let byte i =
+    try digit (unsafe_get b i) lsl 4 + digit (unsafe_get b (i + 1))
+    with E -> raise (Invalid i)
```

Comment:
- You are returning the offset in the resulting string; isn't it more useful/natural to report the location error in the input? (That's how I read the docstring anyway.)  That said, I'm not so sure that returning any location information is that useful in practice.

- In terms of implementation, you could raise the external exception directly in the digit sub-function (passing it the offset).

Reply:
Oops - the offset is wrong. However, I'd be happy to make it return `'a option` too